cpp/src_prims/stats/histogram.h

```diff
@@ -175,21 +175,92 @@ void smemBitsHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
   CUDA_CHECK(cudaGetLastError());
 }
 
+DI void clearHashTable(int2* ht, int hashSize) {
+  for (auto i = threadIdx.x; i < hashSize; i += blockDim.x) {
+    ht[i] = {-1, 0};
+  }
+  __syncthreads();
+}
+
+DI int findEntry(int2* ht, int hashSize, int binId) {
+  int idx = binId % hashSize;
+  int t;
+  int count = 0;
+  while ((t = atomicCAS(&(ht[idx].x), -1, binId)) != -1 && t != binId) {
+    ++count;
+    if (count >= hashSize) {
+      idx = -1;
+      break;
+    }
```

Comment:
This is a sequential search for an empty slot in the HT, right?  And if every entry is looked at and none is found, then the HT is flushed (in the calling code).  I have a feeling this will be very slow, but if you have tests that show it isn't, then please leave it be.  In the worst case, 1024 threads will fill the HT in 5 iterations (5*1024 < 6047).  Perhaps a better solution is to have count be less than some small value, like 30 or 100, so if you have 30 attempts to find a bin and you don't succeed, you flush.

Reply:
Valid point! This count threshold needs to be set to properly balance between the penalty of finding an empty and the cost of flushing the hash table. Let me do that.

Reply:
In the interest of time, I have reduced this threshold to a low value of 10. (meaning 10 tries from a thread, before it declares "we need a flush"). We certainly need to experimentally decide this number.